# [`11. planEvent (Асинхронное выполнение функции)`](../index.md)

Реализуйте функцию `planEvent`, которая планирует вызов функции `cb` через некоторое время(`timeout`) и возвращает `Promise` с результатом выполнения функции `cb`.
В случае `timeout` <= 0 вызов должен произойти сразу же.
Подробнее о `Promise` и `async/await` можно прочесть [здесь](https://learn.javascript.ru/async).

**Ошибки, которые должны быть обработаны:**

1. Первый аргумент не типа `function`. Код ошибки `INVALID_ARGUMENT`.
1. Второй аргумент не типа `number`. Код ошибки `INVALID_ARGUMENT`.

**Примеры использования:**

1. `const f = () => 'Done';`<br/>
   `const result = await planEvent(f, 3000);`<br/>
   `//Через 3 секунды`<br/>
   `console.log(result); // 'Done'`
2. `const result = await planEvent('str', 3000); // ошибка с кодом INVALID_ARGUMENT`
3. `const result = await planEvent(() => 'Done', '3000'); // ошибка с кодом INVALID_ARGUMENT`

## Мое решение

```ts
type Func = () => any;

const planEvent = (func: Func, timeout: number) => {
  if (typeof func !== 'function' || !Number.isFinite(timeout)) {
    throw new Error('INVALID_ARGUMENT');
  }

  const promise = new Promise(resolve => {
    if (timeout <= 0) {
      resolve(func());
    } else {
      setTimeout(() => {
        resolve(func());
      }, timeout);
    }
  });
  return promise;
};

export default planEvent;
```
